### PR TITLE
[BitSail]Supprt bitsail table catalog.

### PR DESCRIPTION
**Pull request by hk-lrzy**
Signed-off-by:

## Pre-Checklist

Note: Please complete **_ALL_** items in the following checklist.

- [x] I have read through the [CONTRIBUTING.md](https://github.com/bytedance/bitsail/blob/master/docs/contributing.md) documentation.
- [x] My code has the necessary comments and documentation (if needed).
- [x] I have added relevant tests.

## Purpose

Some description about what this PR wants to do.

## Approaches

Some description about how this PR achives the purpose. 

## Related Issues

_e.g._ Close #796

## New Behavior (screenshots if needed)

N/A

### CODE REVIEW

File: bitsail-base/src/main/java/com/bytedance/bitsail/base/component/DefaultComponentBuilderLoader.java [hunk 1]
```diff
@@ -45,14 +45,21 @@ public DefaultComponentBuilderLoader(Class<T> clazz) {
...
       loaded = true;
     }
     componentName = StringUtils.lowerCase(componentName);
     if (!components.containsKey(componentName)) {
-      throw new BitSailException(CommonErrorCode.CONFIG_ERROR,
-          String.format("Component %s not in interface %s support until now.", componentName, clazz));
+      if (failOnMiss) {
+        throw new BitSailException(CommonErrorCode.CONFIG_ERROR,
+            String.format("Component %s not in interface %s support until now.", componentName, clazz));
+      }
```
Comment (BlockLiu) [inline bitsail-base/src/main/java/com/bytedance/bitsail/base/component/DefaultComponentBuilderLoader.java:61]: add logs when failOnMiss=true?

File: bitsail-common/src/main/java/com/bytedance/bitsail/common/catalog/TableCatalogErrorCode.java [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+import com.bytedance.bitsail.common.exception.ErrorCode;
+
+public enum TableCatalogErrorCode implements ErrorCode {
+
+  TABLE_CATALOG_TABLE_NOT_EXISTS("Table-Catalog-1", "Catalog table not exists."),
+
+  ;
+  private String code;
+
+  private String description;
```
Comment (BlockLiu) [inline bitsail-common/src/main/java/com/bytedance/bitsail/common/catalog/TableCatalogErrorCode.java:31]: I think we can follow IDEA code style advices, like add 'final' before here.

File: bitsail-common/src/main/java/com/bytedance/bitsail/common/catalog/TableCatalogManager.java [hunk 3]
```diff
@@ -0,0 +1,275 @@
...
+              String.format("Writer table definition %s not exists.", writerTableDefinition));
+        }
+        // try to create table when not exists.
+        writerTableCatalog.createTable(writerTableDefinition, readerCatalogTable);
+      }
+
+      // get writer catalog table.
+      writerCatalogTable = writerTableCatalog.getCatalogTable(writerTableDefinition);
+
+      // gtt base table schema.
```
Comment (BlockLiu) [inline bitsail-common/src/main/java/com/bytedance/bitsail/common/catalog/TableCatalogManager.java]: typo

File: bitsail-common/src/main/java/com/bytedance/bitsail/common/catalog/TableCatalogManager.java [hunk 4]
```diff
@@ -0,0 +1,275 @@
...
+          .build());
+    }
+    return columnInfos;
+  }
+
+  private void alterCatalogSchema(CatalogTableAlterDefinition catalogTableAlterDefinition) {
+    if (!catalogTableAlterDefinition.isNotEmpty()) {
+      return;
+    }
+    if (!tableCatalogAddSync &&
```
Comment (BlockLiu) [inline bitsail-common/src/main/java/com/bytedance/bitsail/common/catalog/TableCatalogManager.java:190]: Why use ! here?

File: bitsail-common/src/main/java/com/bytedance/bitsail/common/catalog/TableCatalogManager.java [hunk 5]
```diff
@@ -0,0 +1,275 @@
...
+    }
+
+    finalCatalogColumns = Lists.newArrayList();
+
+    List<CatalogTableColumn> pendingAddTableColumns = Lists.newArrayList();
+    List<CatalogTableColumn> pendingUpdateTableColumns = Lists.newArrayList();
+    List<CatalogTableColumn> pendingDeleteTableColumns = Lists.newArrayList();
+    for (CatalogTableColumn catalogTableColumn : baseCatalogTableSchema.getColumns()) {
+      String baseCatalogColumnName = StringUtils.lowerCase(catalogTableColumn.getName());
+      if (writePrimaryTableColumnMapping.contains(baseCatalogColumnName)) {
```
Comment (BlockLiu) [inline bitsail-common/src/main/java/com/bytedance/bitsail/common/catalog/TableCatalogManager.java:239]: So our Catalog service is case insensitive? Can this be configured by users?

File: bitsail-connectors/bitsail-connectors-legacy/bitsail-connector-hive/src/main/java/com/bytedance/bitsail/connector/legacy/hive/common/HiveTableCatalogFactory.java [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+      return HiveTableCatalog
+          .builder()
+          .database(database)
+          .table(table)
+          .namespace(null)
+          .hiveConf(HiveConfUtils.fromJsonProperties(
+              connectorConfiguration.get(HiveWriterOptions.HIVE_METASTORE_PROPERTIES)))
+          .build();
+    }
+    throw new UnsupportedOperationException();
```
Comment (BlockLiu) [inline bitsail-connectors/bitsail-connectors-legacy/bitsail-connector-hive/src/main/java/com/bytedance/bitsail/connector/legacy/hive/common/HiveTableCatalogFactory.java:52]: I think we can support hive writer. So add todo here?

File: bitsail-common/src/main/java/com/bytedance/bitsail/common/catalog/TableCatalogManager.java [hunk 7]
```diff
@@ -0,0 +1,287 @@
...
+        throw BitSailException.asBitSailException(TableCatalogErrorCode.TABLE_CATALOG_TABLE_NOT_EXISTS,
+            String.format("Reader table definition %s not exists.", readerTableDefinition));
+      }
+
+      // get reader catalog table.
+      readerCatalogTable = readerTableCatalog.getCatalogTable(readerTableDefinition);
+
+      if (!writerTableCatalog.tableExists(writerTableDefinition)) {
+
+        if (!tableCatalogCreateTableNotExists) {
```
Comment (ayonel) [inline bitsail-common/src/main/java/com/bytedance/bitsail/common/catalog/TableCatalogManager.java:132]: Shall we clean the created table when job was failed finally?

File: bitsail-common/src/main/java/com/bytedance/bitsail/common/catalog/TableCatalogManager.java [hunk 8]
```diff
@@ -0,0 +1,287 @@
...
+    }
+    if (!ignoreTableCatalogAddSync &&
+        CollectionUtils.isNotEmpty(catalogTableAlterDefinition.getPendingAddColumns())) {
+      writerTableCatalog.alterTableColumns(
+          TableOperation.ALTER_COLUMNS_ADD,
+          catalogTableAlterDefinition.getPendingAddColumns()
+      );
+    }
+
+    if (!ignoreTableCatalogDeleteSync &&
```
Comment (ayonel) [inline bitsail-common/src/main/java/com/bytedance/bitsail/common/catalog/TableCatalogManager.java:200]: I recommend replace  the order of Delete and Update.   Add->Update->Delete

**Pull request merged**